.github/workflows/test.yml

```diff
@@ -29,45 +28,93 @@ jobs:
       - name: Upload firmware artifact
         uses: actions/upload-artifact@v4
         with:
-          name: frostsnap-v2-${{ github.run_id }}.elf
+          name: frostsnap-${{ github.sha }}.elf
           path: target/riscv32imc-unknown-none-elf/release/v2
           if-no-files-found: error
 
+  generate-bridge-files:
+    needs: device
+    runs-on: ubuntu-latest
+    steps:
+      - uses: actions/checkout@v3
+      - uses: actions/download-artifact@v4
+        with:
+          name: frostsnap-${{ github.sha }}.elf
+          path: target/riscv32imc-unknown-none-elf/release/
+      - uses: subosito/flutter-action@v2
+        with:
+          channel: "stable"
+      - uses: dtolnay/rust-toolchain@stable
+        with:
+          components: clippy
+      - name: Install dependencies
+        run: sudo apt-get update && sudo apt-get install -y libudev-dev
+      - uses: taiki-e/install-action@just
+      - run: just frostsnapp/install-flutter-deps
+      - run: just frostsnapp/install-rust-deps
+      - run: just gen
+      # Cache the generated files for other jobs
+      - name: Cache generated files
+        uses: actions/cache/save@v3
+        with:
+          path: ./frostsnapp
+          key: bridge-files-${{ github.sha }}
+
+  # Non-device tests
   non-device:
+    needs: generate-bridge-files
     runs-on: ubuntu-latest
     steps:
       - uses: actions/checkout@v3
+      - name: Restore generated files
+        uses: actions/cache/restore@v3
+        with:
+          path: ./frostsnapp
+          key: bridge-files-${{ github.sha }}
+          fail-on-cache-miss: true
       - uses: taiki-e/install-action@just
       - uses: dtolnay/rust-toolchain@stable
         with:
           components: rustfmt, clippy
-      - run: sudo apt-get update
-      - run: sudo apt-get install -y libudev-dev
+      - uses: subosito/flutter-action@v2
+        with:
+          channel: "stable"
+      - uses: actions/download-artifact@v4
+        with:
+          name: frostsnap-${{ github.sha }}.elf
+          path: target/riscv32imc-unknown-none-elf/release/
+      - run: sudo apt-get update && sudo apt-get install -y libudev-dev
       - run: just frostsnapp/install-rust-deps
       - run: just lint-non-device --release
       - run: just test --release --all-features
 
+  # Flutter app build
   app-build:
     name: Flutter app build
     runs-on: ubuntu-latest
-    needs: device
+    needs: generate-bridge-files
     steps:
       - uses: actions/checkout@v3
-      - name: Download firmwre elf
-        uses: actions/download-artifact@v4
+      - name: Restore generated files
+        uses: actions/cache/restore@v3
+        with:
+          path: ./frostsnapp
+          key: bridge-files-${{ github.sha }}
+          fail-on-cache-miss: true
+      - uses: actions/download-artifact@v4
         with:
-          name: frostsnap-v2-${{ github.run_id }}.elf
+          name: frostsnap-${{ github.sha }}.elf
```

Comment:
why is this changing and what does it mean?

Reply:
Discussed, the commit sha is more useful for distinguishing artifacts than runner IDs.